> Tests are not passing because pk is being provided, for this test to pass you either must set pk to none or change the code to return 400 when pk is not an UUID (which current code is not doing)
> 
> ```
> self = <kolibri.core.content.test.test_content_app.ContentNodeAPITestCase testMethod=test_contentnode_tree_bad_pk>
> 
>     def test_contentnode_tree_bad_pk(self):
>         response = self.client.get(
>             reverse(
>                 "kolibri:core:contentnode_tree-detail",
>                 kwargs={"pk": "this is not a UUID"},
>             )
>         )
> >       self.assertEqual(response.status_code, 400)
> E       AssertionError: 404 != 400
> ```

I tried with pk as None but it gives me 404 error 
```py
def test_import_metadata_bad_pk(self):
        response = self.client.get(
            reverse(
                "kolibri:core:importmetadata-detail",
                kwargs={"pk": None},
            )
        )
>       self.assertEqual(response.status_code, 400)
E       AssertionError: 404 != 400
```
```python
kolibri/core/content/test/test_public_api.py:105: AssertionError
----------------------------------------------------------------------------- Captured stderr setup -----------------------------------------------------------------------------
ERROR    2024-11-13 01:19:35,987 No tree cache found for 3 levels and 5 children per level
------------------------------------------------------------------------------ Captured log setup -------------------------------------------------------------------------------
ERROR    kolibri.core.content.test.test_channel_upgrade:test_channel_upgrade.py:110 No tree cache found for 3 levels and 5 children per level
----------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------
INFO     2024-11-13 01:19:36,823 127.0.0.1 - - "GET /api/public/v2/importmetadata/None/" 404 0 "" "unknown"
WARNING  2024-11-13 01:19:36,824 Not Found: /api/public/v2/importmetadata/None/
------------------------------------------------------------------------------- Captured log call -------------------------------------------------------------------------------
WARNING  django.request:log.py:224 Not Found: /api/public/v2/importmetadata/None/
=============================================================================== warnings summary ================================================================================
```

but changing into the retrieve function as `pk == "this is not a UUID"` it worked(gives 400 response) only for ImportMetadataViewset

--- a/kolibri/core/content/api.py
+++ b/kolibri/core/content/api.py
@@ -5,7 +5,6 @@ from base64 import urlsafe_b64decode
 from collections import OrderedDict
 from functools import reduce
 from random import sample
-from uuid import UUID
 
 from django.core.cache import cache
 from django.core.exceptions import ValidationError
@@ -861,13 +860,8 @@ class ContentNodeViewset(InternalContentNodeMixin, RemoteMixin, ReadOnlyValuesVi
     pagination_class = OptionalContentNodePagination
 
     def retrieve(self, request, pk=None):
-
-        try:
-            UUID(pk)
-        except ValueError:
-            return Response(
-                {"error": "Invalid UUID format."}, status=status.HTTP_400_BAD_REQUEST
-            )
+        if pk is None:
+            raise Http404
 
         if self._should_proxy_request(request):
             if self.get_queryset().filter(admin_imported=True, pk=pk).exists():

--- a/kolibri/core/content/api.py
+++ b/kolibri/core/content/api.py
@@ -5,6 +5,7 @@ from base64 import urlsafe_b64decode
 from collections import OrderedDict
 from functools import reduce
 from random import sample
+from uuid import UUID
 
 from django.core.cache import cache
 from django.core.exceptions import ValidationError
@@ -862,6 +863,13 @@ class ContentNodeViewset(InternalContentNodeMixin, RemoteMixin, ReadOnlyValuesVi
     def retrieve(self, request, pk=None):
         if pk is None:
             raise Http404
+
+        try:
+            UUID(pk)
+        except ValueError:
+            return Response(
+                {"error": "Invalid UUID format."}, status=status.HTTP_400_BAD_REQUEST
+            )
 
         if self._should_proxy_request(request):
             if self.get_queryset().filter(admin_imported=True, pk=pk).exists():

--- a/kolibri/core/content/api.py
+++ b/kolibri/core/content/api.py
@@ -5,6 +5,7 @@ from base64 import urlsafe_b64decode
 from collections import OrderedDict
 from functools import reduce
 from random import sample
+from uuid import UUID
 
 from django.core.cache import cache
 from django.core.exceptions import ValidationError
@@ -860,8 +861,14 @@ class ContentNodeViewset(InternalContentNodeMixin, RemoteMixin, ReadOnlyValuesVi
     pagination_class = OptionalContentNodePagination
 
     def retrieve(self, request, pk=None):
-        if pk is None:
-            raise status.HTTP_400_BAD_REQUEST
+
+        try:
+            UUID(pk)
+        except ValueError:
+            return Response(
+                {"error": "Invalid UUID format."}, status=status.HTTP_400_BAD_REQUEST
+            )
+
         if self._should_proxy_request(request):
             if self.get_queryset().filter(admin_imported=True, pk=pk).exists():
                 # Used in the update method for remote request retrieval

--- a/kolibri/core/content/public_api.py
+++ b/kolibri/core/content/public_api.py
@@ -2,6 +2,7 @@ from uuid import UUID
 
 from django.db import connection
 from django.db.models import Q
+from django.http import Http404
 from django.http import HttpResponseBadRequest
 from rest_framework import status
 from rest_framework.response import Response
@@ -51,12 +52,16 @@ class ImportMetadataViewset(GenericViewSet):
         :param pk: id parent node
         :return: an object with keys for each content metadata table and a schema_version key
         """
-        try:
-            UUID(pk)
-        except ValueError:
-            return Response(
-                {"error": "Invalid UUID format."}, status=status.HTTP_400_BAD_REQUEST
-            )
+        if pk is None:
+            raise Http404
+        else:
+            try:
+                UUID(pk)
+            except ValueError:
+                return Response(
+                    {"error": "Invalid UUID format."},
+                    status=status.HTTP_400_BAD_REQUEST,
+                )
 
         content_schema = request.query_params.get(
             "schema_version", self.default_content_schema

--- a/kolibri/core/content/public_api.py
+++ b/kolibri/core/content/public_api.py
@@ -51,8 +51,12 @@ class ImportMetadataViewset(GenericViewSet):
         :param pk: id parent node
         :return: an object with keys for each content metadata table and a schema_version key
         """
-        if pk is None:
-            raise status.HTTP_400_BAD_REQUEST
+        try:
+            UUID(pk)
+        except ValueError:
+            return Response(
+                {"error": "Invalid UUID format."}, status=status.HTTP_400_BAD_REQUEST
+            )
 
         content_schema = request.query_params.get(
             "schema_version", self.default_content_schema

--- a/kolibri/core/content/public_api.py
+++ b/kolibri/core/content/public_api.py
@@ -2,8 +2,8 @@ from uuid import UUID
 
 from django.db import connection
 from django.db.models import Q
-from django.http import Http404
 from django.http import HttpResponseBadRequest
+from django.shortcuts import get_object_or_404
 from rest_framework import status
 from rest_framework.response import Response
 from rest_framework.serializers import Serializer
@@ -52,16 +52,13 @@ class ImportMetadataViewset(GenericViewSet):
         :param pk: id parent node
         :return: an object with keys for each content metadata table and a schema_version key
         """
-        if pk is None:
-            raise Http404
-        else:
-            try:
-                UUID(pk)
-            except ValueError:
-                return Response(
-                    {"error": "Invalid UUID format."},
-                    status=status.HTTP_400_BAD_REQUEST,
-                )
+        try:
+            UUID(pk)
+        except ValueError:
+            return Response(
+                {"error": "Invalid UUID format."},
+                status=status.HTTP_400_BAD_REQUEST,
+            )
 
         content_schema = request.query_params.get(
             "schema_version", self.default_content_schema
@@ -84,7 +81,7 @@ class ImportMetadataViewset(GenericViewSet):
 
         # Get the model for the target node here - we do this so that we trigger a 404 immediately if the node
         # does not exist.
-        node = self.get_object()
+        node = get_object_or_404(models.ContentNode.objects.all(), pk=pk)
 
         nodes = node.get_ancestors(include_self=True)
 

--- a/kolibri/core/content/test/test_content_app.py
+++ b/kolibri/core/content/test/test_content_app.py
@@ -515,6 +515,15 @@ class ContentNodeAPIBase(object):
         )
         self.assertEqual(response.status_code, 404)
 
+    def test_contentnode_tree_bad_pk(self):
+        response = self.client.get(
+            reverse(
+                "kolibri:core:contentnode_tree-detail",
+                kwargs={"pk": "this is not UUID"},
+            )
+        )
+        self.assertEqual(response.status_code, 400)
+
     @unittest.skipIf(
         getattr(settings, "DATABASES")["default"]["ENGINE"]
         == "django.db.backends.postgresql",

--- a/kolibri/core/content/test/test_content_app.py
+++ b/kolibri/core/content/test/test_content_app.py
@@ -506,6 +506,15 @@ class ContentNodeAPIBase(object):
         )
         self.assertEqual(response.status_code, 404)
 
+    def test_contentnode_tree_none_pk(self):
+        response = self.client.get(
+            reverse(
+                "kolibri:core:contentnode_tree-detail",
+                kwargs={"pk": None},
+            )
+        )
+        self.assertEqual(response.status_code, 400)
+
     def test_contentnode_tree_bad_pk(self):
         response = self.client.get(
             reverse(
@@ -514,6 +523,7 @@ class ContentNodeAPIBase(object):
             )
         )
         self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.data["error"], "Invalid UUID format.")
 
     @unittest.skipIf(
         getattr(settings, "DATABASES")["default"]["ENGINE"]

--- a/kolibri/core/content/test/test_content_app.py
+++ b/kolibri/core/content/test/test_content_app.py
@@ -513,17 +513,7 @@ class ContentNodeAPIBase(object):
                 kwargs={"pk": None},
             )
         )
-        self.assertEqual(response.status_code, 400)
-
-    def test_contentnode_tree_bad_pk(self):
-        response = self.client.get(
-            reverse(
-                "kolibri:core:contentnode_tree-detail",
-                kwargs={"pk": "this is not a UUID"},
-            )
-        )
-        self.assertEqual(response.status_code, 400)
-        self.assertEqual(response.data["error"], "Invalid UUID format.")
+        self.assertEqual(response.status_code, 404)
 
     @unittest.skipIf(
         getattr(settings, "DATABASES")["default"]["ENGINE"]

--- a/kolibri/core/content/test/test_public_api.py
+++ b/kolibri/core/content/test/test_public_api.py
@@ -95,6 +95,15 @@ class ImportMetadataTestCase(APITestCase):
         )
         self.assertEqual(response.status_code, 400)
 
+    def test_import_metadata_none_pk(self):
+        response = self.client.get(
+            reverse(
+                "kolibri:core:importmetadata-detail",
+                kwargs={"pk": None},
+            )
+        )
+        self.assertEqual(response.status_code, 400)
+
     def test_import_metadata_bad_pk(self):
         response = self.client.get(
             reverse(
@@ -103,6 +112,7 @@ class ImportMetadataTestCase(APITestCase):
             )
         )
         self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.data["error"], "Invalid UUID format.")
 
     def test_schema_version_just_right(self):
         response = self.client.get(

--- a/kolibri/core/content/test/test_public_api.py
+++ b/kolibri/core/content/test/test_public_api.py
@@ -102,7 +102,7 @@ class ImportMetadataTestCase(APITestCase):
                 kwargs={"pk": None},
             )
         )
-        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.status_code, 404)
 
     def test_import_metadata_bad_pk(self):
         response = self.client.get(

